include/swift/Option/Options.td

```diff
@@ -889,6 +889,15 @@ def sanitize_EQ : CommaJoined<["-"], "sanitize=">,
   Flags<[FrontendOption, NoInteractiveOption]>, MetaVarName<"<check>">,
   HelpText<"Turn on runtime checks for erroneous behavior.">;
 
+def sanitize_recover_EQ
+    : CommaJoined<["-"], "sanitize-recover=">,
+      Flags<[FrontendOption, NoInteractiveOption]>,
+      MetaVarName<"<check>">,
+      HelpText<"Specify which sanitizer runtime checks (see -sanitize=) will "
+               "generate instrumentation that allows error recovery. Listed "
+               "checks should be comma separated. Default behavior is to not "
+               "allow error recovery.">;
```

Comment:
How should I understand this flag compared to e.g. tsan's `halt_on_error=0`?  If I set `sanitize-recover=address` does it unconditionally recover or can I say `ASAN_OPTIONS= halt_on_error=1`? 

You mentioned it only supports asan, but isn't this the default behaviour for tsan?  Or is this somehow different from what tsan is doing?

Reply:
Good questions.

The `halt_on_error` flag is actually not TSan specific. The other sanitizers have it too.

> If I set sanitize-recover=address does it unconditionally recover or can I say ASAN_OPTIONS= halt_on_error=1?

No, it does not unconditionally recover. What it does is cause ASan's instrumentation to allow for error recovery. The runtime `halt_on_error` option is used to decide what to do when ASan detects an issue.

Today, setting `ASAN_OPTIONS=halt_on_error=0` doesn't always work. If you compile **without the `-sanitize-recover=address` option** (equivalent to the current behavior of the swift compiler) then the generated instrumentation doesn't allow for error recovery. What this means is that if you set `ASAN_OPTIONS=halt_on_error=0` at runtime and if an ASan issue is caught via instrumentation then the process will **always** halt regardless of how `halt_on_error` is set. However, if ASan catches an issue via one of its interceptors (e.g. memcpy) then `halt_on_error` runtime option is respected.

With `-sanitize-recover=address` the generated instrumentation allows for error recovery which means that the `halt_on_error` runtime option always does what you would expect.

ASan's default for `halt_on_error` is true which means this issue only effects people who choose to not use the default behavior.

> You mentioned it only supports asan, but isn't this the default behaviour for tsan? Or is this somehow different from what tsan is doing?

`halt_on_error` defaults to false for TSan. The difference here is that TSan's instrumentation pass always allows for error recovery but ASan's instrumentation pass does not.

Reply:
Thanks for the explanation!  Any reason not to accept `-sanitize-recover=thread` then for consistency?

Reply:
Well, the flag wouldn't do anything. The ThreadSanitizer pass doesn't have an option to change the instrumentation so there's nothing we can change in the compilation process. It seems odd to me to have a driver/frontend option that doesn't effect the compilation process.

Given that we're adding a new option I'd prefer to limit the ways in which it can be used.